File: 10.ValidBST/validbst.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node *left, *right;
};

int arr[200000];
int idx = 0;
int n = 0;

/* Build tree from preorder input with -1 as NULL */
struct Node* buildTree() {
    if (idx >= n)
        return NULL;

    int val = arr[idx++];
    if (val == -1)
        return NULL;

    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = val;
    node->left = buildTree();
    node->right = buildTree();

    return node;
}

/* Validate BST using range */
int isValidBST(struct Node* root, int min, int max) {
    if (root == NULL)
        return 1;

    if (root->data <= min || root->data >= max)
        return 0;

    return isValidBST(root->left, min, root->data) &&
           isValidBST(root->right, root->data, max);
}

int main() {
    /* Read input until EOF */
    while (scanf("%d", &arr[n]) != EOF) {
        n++;
    }

    struct Node* root = buildTree();

    /* Value range: -10^5 to 10^5 */
    if (isValidBST(root, -100001, 100001))
        printf("True");
    else
        printf("False");

    return 0;
}

File: 10.ValidBST/validbst.py
# Write your code here!!

import sys
sys.setrecursionlimit(10**7)

class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None


# Read input like Scanner.hasNextInt()
data = list(map(int, sys.stdin.read().split()))
it = iter(data)


def build_tree():
    try:
        val = next(it)
    except StopIteration:
        return None

    if val == -1:
        return None

    node = Node(val)
    node.left = build_tree()
    node.right = build_tree()
    return node


def is_valid_bst(root, min_val, max_val):
    if root is None:
        return True

    if not (min_val < root.data < max_val):
        return False

    return (is_valid_bst(root.left, min_val, root.data) and
            is_valid_bst(root.right, root.data, max_val))


# Build tree
root = build_tree()

# Validate BST using given constraints
print("True" if is_valid_bst(root, -100001, 100001) else "False")
